SEP XXX -- Concrete descriptions of non-canonical DNA, RNA, and proteins
========================================================================

**SEP**               | **XXX**
----------------------|--------------
**Title**             | Concrete descriptions of biopolymers
**Authors**           | [Jonathan Karr](mailto:karr@mssm.edu)
**Editor**            | 
**Type**              | Procedure
**Status**            | Draft
**Created**           | 20-Jun-2019
**Last modified**     | 
**Issue**             | 

## Abstract

To facilitate the understanding, reuse, and composition of genetic designs, SBOL needs to concretely capture the structure of each part. This must include non-canonical nucleic and amino acids that are synthesized via epigenetic, post-transcriptional, and post-translational modification, as well as non-canonical amino acids incorporated via modified translational systems. SBOL should also capture other covalent bonds such as crosslinks in DNA and disulfide bonds. Currently, SBOL does not have a clear mechanism to capture this information. This SEP recommends using the `BpForms` encoding to describe this information within `elements` children of `Sequence` objects.

## Table of Contents

* [1. Rationale](#rationale)
* [2. Specification](#specification)
  * 2.1 2.1 `BpForms` syntax
  * 2.2 `BpForms` alphabets
  * 2.3 Embedding `BpForms` into SBOL 
* [3. Examples](#example)
  * 3.1 Modified RNA: *Bacillus subtilis* tRNA<sup>ILE</sup> 69 
  * 3.2 Modified Protein: Lipoate ligated *Bacillus subtilis* PdhC
* [4. Backwards Compatibility](#compatibility)
* [5. Interpreting `BpForms`: Verifying `BpForms` and calculating their properties](#software)
* [6. Contributing to `BpForms`](#contributing)
* [7. Discussion](#discussion)
  * 7.1 Verbosity of the `BpForms` syntax
  * 7.2 Lack of support for the PSI-MI ontologies of monomeric forms of proteins
  * 7.3 Integrating the `BpForms` software into libSBOL
  * 7.4 Concretely representing macromolecular complexes
* [References](#references)
* [Copyright](#copyright)

## 1. Rationale <a name="rationale"></a>

To facilitate the understanding, reuse, and composition of genetic designs, SBOL needs to concretely capture the structure of each part. This must include non-canonical nucleic and amino acids that are synthesized via epigenetic, post-transcriptional, and post-translational modification, as well as non-canonical amino acids incorporated via modified translational systems. SBOL should also capture other covalent bonds such as crosslinks in DNA and disulfide bonds.

This information is needed for several reasons:
- This information can be used to identify similarities and differences between parts. For example, this information can help disambiguate between the multiple phosphorylated forms of ERK.
- This information can help identify potential interactions between synthetic parts. For example, interactions between multiple phosphorylated forms of ERK and MEK.
- This information can help identify the dependencies of each part that must be included when a part is introduced into another system. For example, PdhC of pyruvate dehydogenase requires a lipoate ligase. This dependence information is conveyed by capturing that PhdC includes a lipoylysine.

## 2. Specification <a name="specification"></a>

This SEP recommends using the `BpForms` string format to concretely capture the primary structure of each DNA, RNA, and protein. `BpForms` is an enhancement over the IUPAC and IUBMB encodings which can describe non-canonical bases (both inline and via alphabets) and crosslinks between non-adjacent monomeric forms. `BpForms` includes three consensus alphabets of hundreds of non-canonical monomeric forms of DNA, RNA, and proteins. Detailed information about the `BpForms` syntax and alphabets is available at https://www.bpforms.org. This includes a formal definition of the `BpForms` grammar.

### 2.1 `BpForms` syntax
`BpForms` uses an IUPAC/IUBMB-like syntax and alphabets to succinctly describe DNA, RNA, and proteins with non-canonical bases. See the [`BpForms` documentation](https://www.bpforms.org) for detailed information and examples.
- Monomeric forms with single-letter codes are indicated by their codes, e.g.
  ```
  ACTGGA
  ```
- Monomeric forms with multi-letter codes are indicated by their codes enclosed in curly brackets, e.g.
  ```
  ACT{m2G}GA
  ```
- Monomeric forms which are not already in the alphabets can be described "inline" in square brackets. These inline monomeric forms can also be used to capture uncertainty about the mass, charge, and position of non-canonical monomeric forms.
  ```
  ACT[id: "AA0305"
    | name: "N5-methyl-L-arginine"
    | structure: "O=C[C@H](CCCN(C(=[NH2])N)C)[NH3+]"
    | backbone-bond-atom: C2
    | backbone-displaced-atom: H2
    | right-bond-atom: C2
    | left-bond-atom: N15-1
    | left-displaced-atom: H15+1
    | left-displaced-atom: H15+1
    ]A
  ```
- Crosslinks can be described using the `crosslink` attribute
  ```
  CAC | crosslink: [
    left-bond-atom: 1S1 |
    left-displaced-atom: 1H1 |
    right-bond-atom: 1S1 |
    right-displaced-atom: 1H1
    ]
  ``` 

### 2.2 `BpForms` alphabets
`BpForms` includes three consensus alphabets of hundreds of non-canonical monomeric forms. These alphabets were built by quality controling and merging the best existing DNA, RNA, and protein alphabets. See the [`BpForms` website](https://www.bpforms.org) for detailed information about the alphabets and the monomeric forms in each alphabet.
- **DNA nucleobases:** The four canonical DNA nucleobases, plus the non-canonical DNA nucleobases in DNAmod
- **RNA nucleosides:** The four canonical RNA nucleosides, plus the non-canonical RNA nucleosides in MODOMICS and the RNA Modification Database
- **Protein residues:** The 20 canonical protein residues, plus the non-canonical protein residues in the PDB Chemical Component Dictionary and RESID

### 2.3 Embedding `BpForms` into SBOL
`BpForms` strings can be embedded within the `elements` attribute of `Sequence` objects. The following URIs should be used to indicate the encodings for the sequences of DNA, RNA, and protein parts.
- DNA: http://edamontology.org/format_3909#dna
- RNA: http://edamontology.org/format_3909#rna
- Protein: http://edamontology.org/format_3909#protein


## 3. Examples <a name='example'></a>

### 3.1 Modified RNA: *Bacillus subtilis* tRNA<sup>ILE</sup> 69 
The following example illustrates how to use `BpForms` to describe the modifications in *Bacillus subtilis* tRNA<sup>ILE</sup> 69 (SynBioHub: [BO_28687](https://synbiohub.org/public/bsu/BO_28687)). The complete SBOL file is available [here](https://github.com/KarrLab/bpforms/blob/master/examples/BO_28687.xml).

```
...
  <sbol:Sequence>
    <sbol:elements>
      GGGCCUGUAGCUCAGC{8U}GG{8U}{8U}AGAGCGCACGCCUGAU{62A}AGCGUGAG{7G}UCGAUGG{5U}{9U}CGAGUCCAUUCAGGCCCACCA
    </sbol:elements>
    <sbol:encoding rdf:resource="http://edamontology.org/format_3909#rna"/>
  </sbol:Sequence>
...
```

### 3.1 Modified Protein: Lipoate ligated *Bacillus subtilis* PdhC

The following example illustrates how to use `BpForms` to describe the lipoate ligation of the acetyltransferase component PdhC of *B. subtilis* pyruvate dehydrogenase complex (SynBioHub: [BO_32431](https://synbiohub.org/public/bsu/BO_32431)). The complete SBOL file is available [here](https://github.com/KarrLab/bpforms/blob/master/examples/BO_32431.xml).

```
...
  <sbol:Sequence>
    <sbol:elements>
      MAFEFKLPDIGEGIHEGEIVKWFVKPNDEVDEDDVLAEVQND{AA0118}AVVEIPSPVKGKVLELKVEEGTVATVGQTIITFDAPGYEDLQFKGSDE
      SDDAKTEAQVQSTAEAGQDVAKEEQAQEPAKATGAGQQDQAEVDPNKRVIAMPSVRKYAREKGVDIRKVTGSGNNGRVVKEDIDSFVNGGAQEAAPQE
      TAAPQETAAKPAAAPAPEGEFPETREKMSGIRKAIAKAMVNSKHTAPHVTLMDEVDVTNLVAHRKQFKQVAADQGIKLTYLPYVVKALTSALKKFPVL
      NTSIDDKTDEVIQKHYFNIGIAADTEKGLLVPVVKNADRKSVFEISDEINGLATKAREGKLAPAEMKGASCTITNIGSAGGQWFTPVINHPEVAILGI
      GRIAEKAIVRDGEIVAAPVLALSLSFDHRMIDGATAQNALNHIKRLLNDPQLILMEA
    </sbol:elements>
    <sbol:encoding rdf:resource="http://edamontology.org/format_3909#protein"/>
  </sbol:Sequence>
...
```

## 4. Backwards Compatibility <a name='compatibility'></a>

Backward can be achieved in multiple ways:
- Parts can have multiple instances of `Sequence,` one with `BpForms` and a second with the IUPAC or IUBMB encoding
- The `BpForms` software can generate IUPAC and IUBMB strings from `BpForms`-encoded strings. This uses information about the parent of each monomeric form (e.g., the parent of selenocysteine is cysteine).
- `BpForms` is also backward compatible with the MODOMICS notation for modified RNA and ProForma notation for modified proteins.

## 5. Interpreting `BpForms`: Verifying `BpForms` and calculating their properties <a name='software'></a>

There are several tools for verifying `BpForms` and calculating properties such as their structure, formula, molecular weights, and charges. See the [`BpForms` website](https://www.bpforms.org) for more information.
- Web validation form
- JSON REST API
- Command line program
- Python API

## 6. Contributing to `BpForms` <a name='contributing'></a>

Contributions to the alphabets and software are welcome via Git pull requests.

## 7. Discussion <a name='discussion'></a>

### 7.1 Verbosity of the `BpForms` syntax
`BpForms` intentionally uses an IUPAC/IUBMB-like syntax, which is verbose for several reasons
- This provides maximal backward compatibility with the IUPAC and IUBMB encodings.
- This enables `BpForms` to directly represent the structures of molecules distinct from the reactions which generate them. This avoids circular definitions of parts, which gives the opportunity to verify parts definitions against biosynthesis reactions.
- This enables `BpForms` to represent any arbitrary part, including parts which are not based on naturally occurring genes.
- This is compatible with the FASTA file format and programs for reading and writing FASTA files.
- This format is also conducive to transcriptomics, proteomics, and systems biology research. In particular, this format is consistent with the MODOMICS syntax for modified RNA and the ProForma syntax for modified proteins. We anticipate that this will encourage broader adoption and compatibility across these fields.

### 7.2 Lack of support for the PSI-MI ontologies of monomeric forms of proteins
PSI-MI is the "standard" ontology for monomeric forms of proteins. However, because PSI-MI lacks concrete definitions of monomeric forms, we cannot integrate PSI-MI with our alphabet, the PDB-CCD, or RESID. To use PSI-MI, we would need to annotate the structure of each term and then link identical structures between the `BpForms` alphabet and PSI-MI.

### 7.3 Integrating the `BpForms` software into libSBOL
To maximize the utility of `BpForms,` the `BpForms` software should be integrated into libSBOL.

### 7.4 Concretely representing macromolecular complexes
To concretely describe every part, we plan to develop another format for concretely describing protein complexes, including the stoichiometry of each subunit and crosslinks between subunits.

## References <a name='references'></a>

[1]: Lang PF, Chebaro Y & Jonathan R. Karr. BpForms: a toolkit for concretely describing modified DNA, RNA, and proteins. [arXiv:1903.10042](https://arxiv.org/abs/1903.10042). [https://www.bpforms.org](https://www.bpforms.org).

## Copyright <a name='copyright'></a>

<p xmlns:dct="http://purl.org/dc/terms/" xmlns:vcard="http://www.w3.org/2001/vcard-rdf/3.0#">
  <a rel="license"
     href="http://creativecommons.org/publicdomain/zero/1.0/">
    <img src="http://i.creativecommons.org/p/zero/1.0/88x31.png" style="border-style: none;" alt="CC0" />
  </a>
  <br />
  To the extent possible under law,
  <a rel="dct:publisher"
     href="sbolstandard.org">
    <span property="dct:title">SBOL developers</span></a>
  has waived all copyright and related or neighboring rights to
  <span property="dct:title">SEP XXX</span>.
This work is published from:
<span property="vcard:Country" datatype="dct:ISO3166"
      content="US" about="sbolstandard.org">
  United States</span>.
</p>
